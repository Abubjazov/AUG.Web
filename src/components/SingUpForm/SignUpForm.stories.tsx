import type { Meta, StoryObj } from '@storybook/react'
import MockedProvider from 'mockData/mockedReduxProvider'

import SignUpForm, { SignUpFormProps } from './SignUpForm'

import '/src/index.css'

const meta: Meta<SignUpFormProps> = {
  component: SignUpForm,
  title: 'COMPONENTS/Sign Up Form',
  parameters: {
    docs: {
      description: {
        component: 'Interface component - SignUpForm',
      },
    },
  },
  tags: ['autodocs'],
  argTypes: {
    userFunction: {
      description: 'Function for "Cancel" button inside form',
    },
    onSignUp: {
      description: 'User registration function for inside "submitHandler"',
    },
  },
}

export default meta

type Story = StoryObj<SignUpFormProps>

export const Default: Story = {
  args: {},

  decorators: [
    Story => (
      <MockedProvider>
        <div
          style={{
            maxWidth: '300px',
            background: '#999999',
            padding: 30,
            borderRadius: 8,
          }}
        >
          <Story />
        </div>
      </MockedProvider>
    ),
  ],
}
