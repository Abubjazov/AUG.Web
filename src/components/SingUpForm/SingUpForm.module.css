.root {
  display: flex;
  flex-direction: column;
  gap: 20px;
  color: var(--white);
  animation: var(--animation-visible);
}

.red-text {
  color: var(--red);
}

.button {
  width: 90px;
  height: 40px;
}

.title {
  font-size: 48px;
  font-weight: 900;
  line-height: 48px;
}

.text {
  font-size: 16px;
  font-weight: 500;
  line-height: 16px;
}

.buttons {
  display: flex;
  gap: 20px;
}
