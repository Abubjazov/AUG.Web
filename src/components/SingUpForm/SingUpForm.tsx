import { FC } from 'react'

import { useInput } from 'hooks/useInput/useInput'
import { useAppSelector } from 'store/hooks'
import { ISignUpData } from 'store/slices/authSlice'
import BaseButton from 'uikit/BaseButton'
import { EBaseButtonMode } from 'uikit/BaseButton/BaseButton'
import BaseInput from 'uikit/BaseInput'

import styles from './SingUpForm.module.css'
export interface SingUpFormProps {
  userFunction: () => void
  onSignUp: (data: ISignUpData) => void
}

const SingUpForm: FC<SingUpFormProps> = ({ userFunction, onSignUp }) => {
  const { isInProgress } = useAppSelector(state => state.auth)

  const email = useInput('', {
    isEmpty: { value: true, message: 'Email address required' },
    isEmail: { value: true, message: 'Please enter a correct email' },
  })

  const password = useInput('', {
    isEmpty: { value: true, message: 'Password required' },
    minLength: { value: 8, message: 'Minimum password length 8 symbols' },
  })

  const confirmPassword = useInput('', {
    isEmpty: {
      value: true,
      message: 'Please confirm your password',
    },
    isValueMatched: {
      value: true,
      comparisonValue: password.value,
      message: 'Password mismatch',
    },
  })

  const submitHandler = () => {
    email.onBlur()
    password.onBlur()
    confirmPassword.onBlur()

    if (
      !email.errors.length &&
      !password.errors.length &&
      !confirmPassword.errors.length
    ) {
      onSignUp({ email: String(email.value), password: String(password.value) })
    }
  }

  return (
    <div className={styles.root}>
      <span className={styles.title}>
        Sign <span className={styles['red-text']}>up</span>
      </span>

      <BaseInput
        dataTestId={'email-input'}
        type={'email'}
        name={'email'}
        placeholder={'email@example.com'}
        value={email.value}
        onChange={email.onChange}
        onBlur={email.onBlur}
        errors={email.errors}
        errorWhite
        isDirty={email.isDirty}
      />

      <BaseInput
        dataTestId={'password-input'}
        type={'password'}
        name={'password'}
        placeholder={'password'}
        value={password.value}
        onChange={password.onChange}
        onBlur={password.onBlur}
        errors={password.errors}
        errorWhite
        isDirty={password.isDirty}
      />

      <BaseInput
        dataTestId={'confirm-password-input'}
        type={'password'}
        name={'confirm-password'}
        placeholder={'password confirm'}
        value={confirmPassword.value}
        onChange={confirmPassword.onChange}
        onBlur={confirmPassword.onBlur}
        errors={confirmPassword.errors}
        errorWhite
        isDirty={confirmPassword.isDirty}
      />

      <div className={styles.buttons}>
        <BaseButton
          userStyles={styles.button}
          label={'Sign up'}
          mode={EBaseButtonMode.CONTAINED_RED}
          onClick={submitHandler}
          loading={isInProgress}
        />

        <BaseButton
          userStyles={styles.button}
          label={'Cancel'}
          onClick={userFunction}
          disabled={isInProgress}
        />
      </div>
    </div>
  )
}

export default SingUpForm
