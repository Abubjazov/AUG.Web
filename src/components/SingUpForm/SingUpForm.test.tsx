import { fireEvent, render, screen } from '@testing-library/react'
import MockedProvider from 'mockData/mockedReduxProvider'

import SingUpForm from './SingUpForm'

describe('SingUpForm', () => {
  test('should render SingUpForm default', () => {
    const mockFn = vi.fn()

    const { asFragment } = render(
      <MockedProvider>
        <SingUpForm userFunction={mockFn} onSignUp={mockFn} />
      </MockedProvider>,
    )

    expect(asFragment()).toMatchSnapshot()
  })

  test('should call SignUn function', () => {
    const mockFn = vi.fn()

    render(
      <MockedProvider>
        <SingUpForm userFunction={mockFn} onSignUp={mockFn} />
      </MockedProvider>,
    )

    fireEvent.click(screen.getByText('Sign up'))

    expect(mockFn).toHaveBeenCalledTimes(0)

    fireEvent.change(screen.getByPlaceholderText('email@example.com'), {
      target: { value: 'email@test.tst' },
    })

    fireEvent.change(screen.getByPlaceholderText('password'), {
      target: { value: '12345678' },
    })

    fireEvent.change(screen.getByPlaceholderText('password confirm'), {
      target: { value: '12345678' },
    })

    fireEvent.click(screen.getByText('Sign up'))

    expect(mockFn).toHaveBeenCalledTimes(1)
  })
})
