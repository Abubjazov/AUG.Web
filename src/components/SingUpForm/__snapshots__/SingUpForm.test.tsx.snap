// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SingUpForm > should render SingUpForm default 1`] = `
<DocumentFragment>
  <div
    class="_root_702078"
  >
    <span
      class="_title_702078"
    >
      Sign 
      <span
        class="_red-text_702078"
      >
        up
      </span>
    </span>
    <div
      class="_root_672eb8"
    >
      <input
        data-testid="email-input"
        name="email"
        placeholder="email@example.com"
        type="email"
        value=""
      />
    </div>
    <div
      class="_root_672eb8"
    >
      <input
        data-testid="password-input"
        name="password"
        placeholder="password"
        type="password"
        value=""
      />
    </div>
    <div
      class="_root_672eb8"
    >
      <input
        data-testid="confirm-password-input"
        name="confirm-password"
        placeholder="password confirm"
        type="password"
        value=""
      />
    </div>
    <div
      class="_buttons_702078"
    >
      <button
        class="_root_91d6ec _contained-red_91d6ec  _button_702078"
        data-testid="base-button"
        type="button"
      >
        Sign up
      </button>
      <button
        class="_root_91d6ec _outlined-white_91d6ec  _button_702078"
        data-testid="base-button"
        type="button"
      >
        Cancel
      </button>
    </div>
  </div>
</DocumentFragment>
`;
